protoc-gen-twirp_php/templates/service/JsonClient.php

```diff
@@ -0,0 +1,270 @@
+<?php
+# Generated by the protocol buffer compiler (protoc-gen-twirp_php {{ .Version }}).  DO NOT EDIT!
+# source: {{ .File.Name }}
+
+declare(strict_types=1);
+
+namespace {{ .File | phpNamespace }};
+
+use Google\Protobuf\Internal\GPBDecodeException;
+use Google\Protobuf\Internal\Message;
+use Http\Discovery\Psr17FactoryDiscovery;
+use Http\Discovery\Psr18ClientDiscovery;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Twirp\Context;
+use Twirp\Error;
+use Twirp\ErrorCode;
+
+/**
+ * A JSON client that implements the {@see {{ .Service | phpServiceName .File }}} interface.
+ * It communicates using JSON and can be configured with a custom HTTP Client.
+ *
+ * Generated from protobuf service <code>{{ .Service | protoFullName .File }}</code>
+ */
+final class {{ .Service | phpServiceName .File }}JsonClient implements {{ .Service | phpServiceName .File }}
```

Comment:
The diff between this file and Client is [this](https://gist.github.com/jitendra-1217/76ec69ca196b272995cdede1d05588cc). I was thinking of but not sure of putting a abstract class.

Reply:
I don't think an abstract class is necessary here. It's just generated code. Also, if one day we see that fit, we can also change the code in a backwards compatible way.

I'm rather thinking about a common template though. Most of the template is the same, apart from a few places.

WDYT?

Reply:
Agree to have a common template for client code. I would check.

Reply:
Hey @sagikazarmark , sorry about bit late reply here.

I tried it. In total I found it bit tricky wrt scope of change in generation code(attached), and template files itself- makes it difficult to read.

I was thinking if it is possible to take current change as in?

<details>
<summary>Tentative changes (does not include template changes)</summary>

```diff
diff --git a/protoc-gen-twirp_php/internal/gen/generator.go b/protoc-gen-twirp_php/internal/gen/generator.go
index 7bedb66..d8b500d 100644
--- a/protoc-gen-twirp_php/internal/gen/generator.go
+++ b/protoc-gen-twirp_php/internal/gen/generator.go
@@ -29,7 +29,7 @@ type Request struct {
 	GlobalFiles []string
 
 	// ServiceFiles are generated once per service with the service name as a prefix.
-	ServiceFiles []string
+	ServiceFiles []TemplateFile
 
 	// Version is the compiler's version.
 	Version string
@@ -40,6 +40,11 @@ func New(box packr.Box) Generator {
 	return &generator{box}
 }
 
+type TemplateFile struct {
+	Path string
+	LayoutPath string
+}
+
 type generator struct {
 	box packr.Box
 }
@@ -100,7 +105,7 @@ func (g *generator) generateServiceFile(
 	ctx *generatorContext,
 	file *descriptor.FileDescriptorProto,
 	svc *descriptor.ServiceDescriptorProto,
-	serviceFile string,
+	serviceFile TemplateFile,
 ) (*plugin.CodeGeneratorResponse_File, error) {
 	data := &serviceFileData{
 		File:         file,
@@ -109,12 +114,17 @@ func (g *generator) generateServiceFile(
 		Version:      ctx.request.Version,
 	}
 
-	tpl, err := g.box.MustString(serviceFile)
+	tpl, err := g.box.MustString(serviceFile.Path)
+	if err != nil {
+		return nil, err
+	}
+	layout, err := g.box.MustString(serviceFile.LayoutPath)
 	if err != nil {
 		return nil, err
 	}
+	tpl += layout
 
-	tpl, err = executeTemplate(ctx, tpl, data)
+	tpl, err = executeTemplate(ctx, tpl, data, "")
 	if err != nil {
 		return nil, err
 	}
@@ -124,7 +134,7 @@ func (g *generator) generateServiceFile(
 			"%s/%s%s",
 			php.Path(file),
 			php.ServiceName(file, svc),
-			strings.Replace(path.Base(serviceFile), "_Service_", "", -1),
+			strings.Replace(path.Base(serviceFile.Path), "_Service_", "", -1),
 		)),
 		Content: proto.String(tpl),
 	}, nil
diff --git a/protoc-gen-twirp_php/internal/gen/template.go b/protoc-gen-twirp_php/internal/gen/template.go
index b733e25..028ee96 100644
--- a/protoc-gen-twirp_php/internal/gen/template.go
+++ b/protoc-gen-twirp_php/internal/gen/template.go
@@ -32,7 +32,7 @@ func TxtFuncMap(ctx *generatorContext) template.FuncMap {
 }
 
 // executeTemplate executes a template and returns the result.
-func executeTemplate(ctx *generatorContext, t string, data interface{}) (string, error) {
+func executeTemplate(ctx *generatorContext, t string, data interface{}, tname string) (string, error) {
 	buf := new(bytes.Buffer)
 
 	tpl, err := template.New("").Funcs(TxtFuncMap(ctx)).Parse(t)
diff --git a/protoc-gen-twirp_php/main.go b/protoc-gen-twirp_php/main.go
index 396ed94..f380bae 100644
--- a/protoc-gen-twirp_php/main.go
+++ b/protoc-gen-twirp_php/main.go
@@ -52,10 +52,11 @@ func Main(in io.Reader, out io.Writer, box packr.Box) error {
 		GlobalFiles: []string{
 			"global/TwirpError.php",
 		},
-		ServiceFiles: []string{
-			"service/_Service_.php",
-			"service/Client.php",
-			"service/Server.php",
+		ServiceFiles: []gen.TemplateFile{
+			gen.TemplateFile{"service/_Service_.php", ""},
+			gen.TemplateFile{"service/Client.php", "layout/Client.php"},
+			gen.TemplateFile{"service/JsonClient.php", "layout/Client.php"},
+			gen.TemplateFile{"service/Server.php", ""},
 		},
 		Version: version,
 	}
```
</details>